src/NuGet.Clients/NuGet.VisualStudio.Common/WindowFrameHelper.cs

```diff
@@ -23,5 +23,11 @@ public static void AddF1HelpKeyword(IVsWindowFrame windowFrame, string keywordVa
                 }
             }
         }
+
+        public static void DisableWindowAutoReopen(IVsWindowFrame windowFrame)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+            ErrorHandler.ThrowOnFailure(windowFrame.SetProperty((int)__VSFPROPID5.VSFPROPID_DontAutoOpen, true));
```

Comment:
Don't throw an exception. We don't want to hang/crash VS because of an error.

If the windowFrame fails to set the property, that's an indication of a serious problem that should be handled somewhere else.

What NuGet PM UI/Console needs to know to handle that failed operation?

Reply:
Good point. However, it wouldn't hang/crash....

Here's what would have happened: The window would not have opened, no error reported, and if you attempted to re-open the window, it would open (but not have this property set).

@dominoFire  what do you think about this change?
Rather than destroy the Window if a property fails, I propose to emit telemetry so we can track when properties fail to set. I know it's probably unlikely to happen, but it's probably a good habit for us to get into.

@zivkan is in-progress with an Exception Emit method, so I copied that from here: https://github.com/NuGet/NuGet.Client/pull/3249/commits/4835b1fcc5477e49ca7f609ba0e80e3831e15a90

Reply:
What are we going to do about it if it fails and we get the telemetry? 

We should not add too many granular try catches.
The calling code of this method itself throws, so I would discourage against adding a try/catch statement here, but if we think it's necessary, somewhere higher on the stack. 

The telemetry should also be used for "controlled" failures, and I'm not sure this is one of those.

Reply:
> The calling code throws

The calling code you're referring to is this?https://github.com/NuGet/NuGet.Client/blob/f3581c38ef62b2e4f4a6e299a712c64562147e73/src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs#L527 
Here we will throw if **creating** the Window had a problem. 
This change is only to modify a property of that Window as a separate action. I believe this doesn't capture errors from that second action.

Granular: I only need the try/catch because the Emit method requires an Exception. I could just digest the boolean, if you like that?

What we do with telemetry in general is out of scope for this PR.
If this isn't a "controlled" failure, and we shouldn't report telemetry, then what are you proposing? Just silently ignore the error?

dominofire said:
> If the windowFrame fails to set the property, that's an indication of a serious problem that should be handled somewhere else.

Answer: it's not going to be handled anywhere.

Reply:
> Here we will throw if creating the Window had a problem.
This change is only to modify a property of that Window as a separate action. I believe this doesn't capture errors from that second action.

I was just pointing out the inconsistency. 
It doesn't need to capture it, I'm just saying they are both throwing, and we don't really do anything with either.

> Granular: I only need the try/catch because the Emit method requires an Exception. I could just digest the boolean, if you like that?

My bad, I missed that fact that you were rethrowing. 

> What we do with telemetry in general is out of scope for this PR.
If this isn't a "controlled" failure, and we shouldn't report telemetry, then what are you proposing? Just silently ignore the error?

@zivkan's just adding that in his PR, and you've copied his new code, so it is in scope. :) 
Anyway don't want to be nitpicky (I'm not gonna block this PR on this discussion), but my primary point is that I don't see this telemetry adding value for us specifically. 

Would it be nice for us to report to the component if we see frequent failures?
Sure, maybe we even report a bug with a component or something, but it sets a precedent of firing off telemetry in every native call that we make. 
We do a lot of that. 

Just consider `how relevant` to the big picture this event is. 
The message exception is an HResult, so we might not learn much there :)

Just se

Reply:
Based on discussion, I've removed the Throw and also reversed the attempt at writing telemetry.
Any error from SetProperty will be silent, and I've added a comment explaining that.

Reply:
> Sure, maybe we even report a bug with a component or something, but it sets a precedent of firing off telemetry in every native call that we make.
We do a lot of that.

I don't understand what you're advocating. I see 3 options:

1. ignore errors from native components, and if customers are experiencing issues, we have no idea. Even if a customer reports an issue, we'll probably be unable to repro locally, so we might still be unable to get enough information to fix it.
2. increase development time by adding in extra error handling that we expect never to get, but be more confident that if our assumptions were wrong, that the customer will still get a decent experience.
3. add telemetry for error cases we don't expect to happen, so if they start happening and we notice the telemetry, we can re-evaluate our assumptions.

Replace "from native components" with "other unexpected errors" and this is true for both Donnie's PR and mine. They're both error conditions we currently expect should NEVER happen. Adding telemetry takes little effort, but will give us some information if we're wrong about it being an error condition that can never happen.

> Just consider how relevant to the big picture this event is.

If our assumptions are wrong about this being an "impossible" error, or if it works now but starts failing because a partner team changed the APIs which we call, causing customers to have a bad experience, it gives us a way to estimate the impact of the bad experience, which we can use to priotitise working on an improvement.

That seems relevant enough to me.

Reply:
I'm not advocating for either of the 3. 
There's a balance between defensive programming and a validation obsession, which I am fully aware is an over-generalization and an exaggeration of what we are doing here. 

If we want these validations we should do them on a higher level, a user action level rather than in very specific places. 

I just don't think our code should be littered with try/catches and if checks everywhere. 
So not against these checks, just specific emphasis on " a higher level, a user action level". 

Let's explore what's happening in the current code:

The newly added method is called from:

[CreateNewWindowFrameAsync](https://github.com/NuGet/NuGet.Client/blob/81239a72b438f65fab0dd761f9f5c85be04830e4/src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs#L384)
and 
[CreateDocWindowForSolutionAsync](https://github.com/NuGet/NuGet.Client/blob/81239a72b438f65fab0dd761f9f5c85be04830e4/src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs#L687)

Each of those is called by one of the following: 

[ShowManageLibraryPackageForSolutionDialog](https://github.com/NuGet/NuGet.Client/blob/81239a72b438f65fab0dd761f9f5c85be04830e4/src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs#L780)

 [ShowManageLibraryPackageDialog](https://github.com/NuGet/NuGet.Client/blob/81239a72b438f65fab0dd761f9f5c85be04830e4/src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs#L584)

Each one of those is a file and forget which fires off a telemetry event...which is the exact same thing what we would've done here.

Reply:
> I just don't think our code should be littered with try/catches

I'd like to reiterate that my code was not littered everywhere, but I created a SetProperty Utility method that could be used by any UI that wants to set any property and report problems with it. 
You also say "higher level", so I'm not sure if you're saying Utility methods are not the right place.

> Each one of those is a file and forget which fires off a telemetry event...which is the exact same thing what we would've done here.

I don't understand what the point is here. Yes, PMUI fires telemetry events. What does that have to do with telemetry for errors?

Reply:
> Each one of those is a file and forget which fires off a telemetry event...which is the exact same thing what we would've done here.

We for native APIs that return an HRESULT instead of throwing exceptions, we should check and throw, as Donnie did in an earlier commit?

That way the file and forget call will get an error, and it won't be silently ignored here.

Reply:
> I'd like to reiterate that my code was not littered everywhere, but I created a SetProperty Utility method that could be used by any UI that wants to set any property and report problems with it.

I am aware of that, so that's why I said `which I am fully aware is an over-generalization and an exaggeration of what we are doing here` before I over-generalized :) 

> You also say "higher level", so I'm not sure if you're saying Utility methods are not the right place.

I will refer to `a user action level rather than in very specific places`. 
Don't want to call out a utility level. Just an action level. This utility is called for a reason. Handle it there. 

> I don't understand what the point is here. Yes, PMUI fires telemetry events. What does that have to do with telemetry for errors?

My bad, I could've provided more context on that rather than jump to the conclusion. 
Refer to https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.vstasklibraryhelper.fileandforget?view=visualstudiosdk-2019

https://github.com/microsoft/vs-threading/blob/master/doc/cookbook_vs.md#void-returning-fire-and-forget-methods

Specifically, the telemetry is fired off when a fault event happens. 
More specifically if you code throws, it is gonna end up in a telemetry event that is already happening.

Reply:
@zivkan 
GitHub didn't update so didn't realize you had replied. 

> So for native APIs that return an HRESULT instead of throwing exceptions, we should check and throw, as Donnie did in an earlier commit?
That way the file and forget call will get an error, and it won't be silently ignored here.

Yes, I think it's good if the code has ThrowOnFailure, the error will get propagated in our telemetry.

Reply:
Long story short: I should go back to how I originally had it. Got it 🤣

Reply:
I originally asked not throw because in other parts of NuGet, we don't throw:

https://github.com/NuGet/NuGet.Client/blob/d902cfba4ad14962316082e5424d4d2d45fd9baf/src/NuGet.Clients/NuGet.Console/WpfConsole/WpfConsole.cs#L322-L323

I'm OK accepting either throw or not to throw.

Reply:
I just forced a thrown exception, and it was recorded in our telemetry under this label:
`vs/nuget/fileandforget/nugetpackage/showmanagelibrarypackagedialog`

I'll consider this PR unblocked. Here's a new issue to track discussion on whether we should handle Exceptions more strictly, and not just let the PMUI exist in partial states.